File: perles.js
// Verification validité string

function verifier_sequence(sequence_a, sequence_b)
{
	var maxI = sequence_a.length;
	for (var i = 0; i < maxI; i++)
	{
		if (sequence_a[i] != sequence_b[i])
			return true;
	}
	return false;
}

function verifier_a_partir_de(morceau)
{
	var maxJ = morceau.length;
	for ( var j = 2; j <= maxJ; j += 2 )
	{
		var sequence_a = morceau.substring(0, j / 2);
		var sequence_b = morceau.substring(j / 2, j);
		if ( !verifier_sequence(sequence_a, sequence_b) )
			return false;
	}
	return true;
}

function verifier(fil) {
	var maxI = fil.length;
	for ( var i = 0; i < maxI; i++ ) {
		if ( !verifier_a_partir_de(fil.substring(i, fil.length)) )
			return false;
	}
	return true;
}

// Verification validité string optimisé (considere que la chaine est vraie si on lui retire la derniere perle)

function verifier_opti(fil) {
	var maxJ = fil.length;
	for ( var j = 2; j <= maxJ; j += 2)
	{
		var sequence_a = fil.substring(maxJ - j, maxJ - j / 2);
		var sequence_b = fil.substring(maxJ - j / 2, maxJ);
		if ( !verifier_sequence(sequence_a, sequence_b) )
			return false;
	}
	return true;
}


// Backtracking

function time_machine(fil)
{
	if ( fil[fil.length - 1] == "0" )
	{
		fil = fil.substring(0, fil.length - 1) + "1";
	}
	else if ( fil[fil.length - 1] == "1" )
		fil = fil.substring(0, fil.length - 1) + "2";
	else
		fil = time_machine(fil.substr(0, fil.length - 1));
	return fil;
}

function ajouter_perle(fil) {
	// avoid stack overflow or too long execution
	if (iterations-- < 0)
		return (fil);

	if (verifier_ft(fil + "0")) {
		return ajouter_perle(fil + "0");
	}
	else if (verifier_ft(fil + "1")) {
		return ajouter_perle(fil + "1");
	}
	else if (verifier_ft(fil + "2")) {
		return ajouter_perle(fil + "2");
	}

	do {
		fil = time_machine(fil);
	} while (!verifier_ft(fil))

	return ajouter_perle(fil);
}



// ####### unit test ########

function test(fil, result)
{
	result_ = verifier_ft(fil);
	 if ( result_ !== result )
		console.log("ERROR :" + fil + " expected " + result + ": returned " + result_);
	else
		console.log("test success");
}

function test_verifier()
{
	var fil;

	fil = "00";
	res = false;
	test(fil, res);

	fil = "01";
	res = true;
	test(fil, res);

	fil = "001";
	res = false;
	test(fil, res);

	fil = "120";
	res = true;
	test(fil, res);

	fil = "12002";
	res = false;
	test(fil, res);

	fil = "01202";
	res = true;
	test(fil, res);

	fil = "012102012";
	res = true;
	test(fil, res);

	fil = "01020120210120102012021020102101201020120210120102101202102010210120102012021012010201";
	res = true;
	test(fil, res);

	fil = "01020120210120102012021020102101201020120210120102101202102010210120102012021012102010210";
	res = true;
	test (fil, res);
}

var iterations = 30000;
var verifier_ft = verifier_opti

function start () {
	test_verifier();

	console.log("iterations : " + iterations);
	var result = ajouter_perle("");
	console.log("RESULT     : " + result);
	console.log("Validity   : " + verifier(result));
}

start ();

File: perles.c
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <stdio.h>

// Utils

int iterations = 30000;

// ############# verifications #############

// 1 = sequence valide | 0 = sequence invalide
char verifier_sequence(char *str, int deb, int mi)
{
	int max = mi - deb;

	for (int i = 0; i < max; i++)
	{
		if (str[deb + i] != str[mi + i])
			return (1);
	}
	return (0);
}

// 1 = sequence valide | 0 = sequence invalide
char verifier(char *fil)
{
	int max = strlen(fil);

	if (max <= 1)
		return (1);
	for (int i = 2; i <= max; i += 2)
		if (!verifier_sequence(fil, max - i, max - (i / 2)))
			return (0);
	return (1);
}

// Backtracking

char *time_machine(char *fil)
{
	if (fil[strlen(fil) - 1] != '2')
		fil[strlen(fil) - 1]++;
	else
	{
		fil[strlen(fil) - 1] = '\0';
		fil = time_machine(fil);
	}
	return (fil);
}

char *ajouter_perle(char *fil)
{
	if (iterations-- < 0)
		return (fil);

	strcat(fil, "0");
	if (verifier(fil))
		return ajouter_perle(fil);
	fil[strlen(fil) - 1] = '\0';

	strcat(fil, "1");
	if (verifier(fil))
		return ajouter_perle(fil);
	fil[strlen(fil) - 1] = '\0';

	strcat(fil, "2");
	if (verifier(fil))
		return ajouter_perle(fil);
	fil[strlen(fil) - 1] = '\0';

	do {
		fil = time_machine(fil);
	} while (!verifier(fil));

	return ajouter_perle(fil);
}

// Main

void 	test(char *str, char value)
{
	int value_ = verifier(str);

	if (value != value_)
		printf("Fail : %s => %d | %d\n", str, value, value_);
	printf("Yay !\n");
}

void 	test_verifier(void)
{
	char *str = "00";
	char value = 0;

	printf("=== Start test ===\n");
	test(str, value);
	str = "0001";
	value = 1;
	test(str, value);
	str = "0001";
	value = 1;
	test(str, value);
	str = "020100";
	value = 0;
	test(str, value);
	str = "020101";
	value = 0;
	test(str, value);
	str = "121121";
	value = 0;
	test(str, value);
	str = "0201210";
	value = 1;
	test(str, value);
	printf("=== End test ===\n");
}

int 	main(int argc, char **argv)
{
	char *str = malloc(iterations);

	if (argc == 2)
		iterations = atoi(argv[1]);
	str[0] = '\0';
	test_verifier();
	ajouter_perle(str);
	write(1, str, strlen(str));
	printf("\n %d", strlen(str));
}